ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/SnapServer.java

```diff
@@ -14,66 +14,616 @@
+                  if (isPartialRange || !statefulPredicate.shouldGetMore()) {
+                    // send a proof for the left side range origin
+                    proofNodes.addAll(
+                        worldStateProof.getStorageProofRelatedNodes(
+                            getAccountStorageRoot(forAccountHash, storage),
+                            forAccountHash,
+                            Hash.wrap(startKeyBytes)));
+                    if (!accountStorages.isEmpty()) {
+                      // send a proof for the last key on the right
+                      proofNodes.addAll(
+                          worldStateProof.getStorageProofRelatedNodes(
+                              getAccountStorageRoot(forAccountHash, storage),
+                              forAccountHash,
+                              Hash.wrap(accountStorages.lastKey())));
+                    }
+                  }
+
+                  if (!statefulPredicate.shouldGetMore()) {
+                    break;
+                  }
+                }
+
+                var resp = StorageRangeMessage.create(collectedStorages, proofNodes);
+                LOGGER.debug(
+                    "returned in {} storage {} to {} range {} to {} with {} storages and {} proofs, resp size {} of max {}",
+                    stopWatch,
+                    asLogHash(range.hashes().first()),
+                    asLogHash(range.hashes().last()),
+                    asLogHash(range.startKeyHash()),
+                    asLogHash(range.endKeyHash()),
+                    collectedStorages.size(),
+                    proofNodes.size(),
+                    resp.getSize(),
+                    maxResponseBytes);
+                return resp;
+              })
+          .orElseGet(
+              () -> {
+                LOGGER.debug("returned empty storage range due to missing worldstate");
+                return EMPTY_STORAGE_RANGE;
+              });
+    } catch (Exception ex) {
+      LOGGER.error("Unexpected exception serving storage range request", ex);
+      return EMPTY_STORAGE_RANGE;
+    }
+  }
+
+  MessageData constructGetBytecodesResponse(final MessageData message) {
+    if (!isStarted.get()) {
+      return EMPTY_BYTE_CODES_MESSAGE;
+    }
+    StopWatch stopWatch = StopWatch.createStarted();
+
+    final GetByteCodesMessage getByteCodesMessage = GetByteCodesMessage.readFrom(message);
+    final GetByteCodesMessage.CodeHashes codeHashes = getByteCodesMessage.codeHashes(true);
+    final int maxResponseBytes = Math.min(codeHashes.responseBytes().intValue(), MAX_RESPONSE_SIZE);
+    LOGGER
+        .atTrace()
+        .setMessage("Receive get bytecodes message for {} hashes")
+        .addArgument(codeHashes.hashes()::size)
+        .log();
+
+    try {
+      List<Bytes> codeBytes = new ArrayDeque<>();
+      for (Bytes32 codeHash : codeHashes.hashes()) {
+        Optional<Bytes> optCode = worldStateStorageCoordinator.getCode(Hash.wrap(codeHash), null);
```

Comment:
In case a peer is asking an empty code hash we can directly return Bytes.EMPTY instead of reading database for nothing

Reply:
I also think we can limit the number of code requests because a node could ask a huge number of empty code hash and spaming our node

Reply:
If we adopt your suggestion about Bytes.EMPTY return for empty code hash, then the response size limit will be decent DoS'd protection.

It seems like p2p rate limiting should be handled at a lower level since we could be dos'd by regular eth messages also 🤔

Reply:
If the answer is empty we will never reach the max limit. Imagine if I ask you for 3000 Empty Code Hash. How can the size limit be reached?